pam_faillock: use vendor specific faillock.conf as fallback

schubi2 opened the pull request:
Use the vendor directory (/usr/etc) as fallback for a distribution provided default config if there is no configuration in /etc.

- pam_faillock.c, faillock.h : Take care about the fallback configuration in /usr/etc.
- pam_faillock.8.xml:  added /usr/etc description

t8m commented on modules/pam_faillock/pam_faillock.8.xml:
```diff
@@ -134,10 +134,15 @@
+		 <para condition="with_vendordir">
+                   Use another configuration file instead of the default
+                   <filename>/etc/security/faillock.conf</filename> or
+		   <filename>%vendordir%/security/faillock.conf</filename> if it does not exists.
```
`instead of the default which is to use the file <filename>/etc/security/faillock.conf</filename> or, if that one is not present, the file <filename>%vendordir%/security/faillock.conf</filename>.`

t8m commented on modules/pam_faillock/pam_faillock.8.xml:
```diff
@@ -328,6 +333,15 @@ session  required       pam_selinux.so open
+        <term><filename>%vendordir%/security/faillock.conf</filename></term>
+        <listitem>
+          <para>
+	    the config file for pam_faillock options. It will be taken if
```
s/taken/used/

schubi2 commented on modules/pam_faillock/tests/faillock.conf.in:
```diff
@@ -0,0 +1,62 @@
+# Configuration for locking the user after multiple failed
+# authentication attempts.
```
It will be used in tst-pam_faillock-retval.c. 
* Testing if loading of a special config file is working at all.
* Decreasing fail attempts and timeout value.

While building faillock.conf will be generated from faillock.conf.in automatically.
Is this way wrong or is there any better one ?

ldv-alt commented:
On Tue, Jan 18, 2022 at 12:55:52AM -0800, Stefan Schubert wrote:
> While building faillock.conf will be generated from faillock.conf.in automatically.
> Is this way wrong or is there any better one ?

Is the generated files always the same as its pattern?
If yes, why one may want to generate this file in the first place?

schubi2 commented:
> On Tue, Jan 18, 2022 at 12:55:52AM -0800, Stefan Schubert wrote: While building faillock.conf will be generated from faillock.conf.in automatically. Is this way wrong or is there any better one ?
> Is the generated files always the same as its pattern? If yes, why one may want to generate this file in the first place?
To be honest it has been the only possible way while calling the build process. I am not so familiar with it. But let me check it again......Sure, I do not like the template too.

ldv-alt commented:
Stefan, _please_ make sure your changes pass the CI.

t8m commented:
We do not allow merge commits. Could you please rebase the PR branch against the current master branch instead?

schubi2 commented:
> We do not allow merge commits. Could you please rebase the PR branch against the current master branch instead?

I hope, I have done the right way ? :-)

t8m commented:
You need to drop the last commit.

ldv-alt commented:
Stefan, _please_ make sure your changes do not contain merge commits.
In other words, please _rebase_ on top of the master branch.

t8m commented:
You've added a merge commit again.

You need to do `git rebase master`, possibly fixing any conflicts and then git force push.

schubi2 commented:
> You've added a merge commit again.
> 
> You need to do `git rebase master`, possibly fixing any conflicts and then git force push.

Yes, I have forgotten to merge the official master with my local master. 
Thanks for the help !

ldv-alt commented on modules/pam_faillock/pam_faillock.c:
```diff
@@ -192,6 +192,13 @@ read_config_file(pam_handle_t *pamh, struct options *opts, const char *cfgfile)
 	char linebuf[FAILLOCK_CONF_MAX_LINELEN+1];
 
 	f = fopen(cfgfile, "r");
+#ifdef VENDOR_FAILLOCK_DEFAULT_CONF
+	if (f == NULL && errno == ENOENT && cfgfile == default_faillock_conf) {
+	  /* If the specified default file in /etc does not exist,
+             try the vendor file as fallback. */
+	  f = fopen(VENDOR_FAILLOCK_DEFAULT_CONF, "r");
+	}
+#endif
 	if (f == NULL) {
 		/* ignore non-existent default config file */
 		if (errno == ENOENT && cfgfile == default_faillock_conf)
```
Before this change, `read_config_file` used to return `PAM_SUCCESS` when `fopen(default_faillock_conf, "r")` failed with `ENOENT`.
After this change, `read_config_file` would return `PAM_SERVICE_ERR` when attempts of opening `default_faillock_conf` and `VENDOR_FAILLOCK_DEFAULT_CONF` are failed with `ENOENT`.

I bet this change of behaviour was not intended.

ldv-alt commented:
I haven't seen tst-pam_faillock-retval.c yet, changes to other files look good.

ldv-alt commented on modules/pam_faillock/tst-pam_faillock-retval.c:
```diff
@@ -0,0 +1,115 @@
+	/* root has access */
+	ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+	ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+		  "auth [success=1 default=bad] %s/../pam_debug/.libs/pam_debug.so auth=success cred=success\n"
+		  "auth [default=die] %s/.libs/%s.so dir=%s authfail even_deny_root conf=%s\n"
+	          "auth sufficient %s/.libs/%s.so dir=%s authsucc even_deny_root conf=%s\n",
```
This looks too complicated.
Why do you need all these jumps if `auth pam_debug.so auth=success cred=success` always succeeds?
Wouldn't something more simple work as well, e.g.
```
auth required %s/../pam_permit/.libs/pam_permit.so
auth required %s/.libs/%s.so authsucc even_deny_root dir=%s conf=%s
```
?

Also, I suggest testing (at least once) all module types and all 6 functions, not just `pam_authenticate`.
You can use tst-pam_sepermit-retval.c as an example.

ldv-alt commented on modules/pam_faillock/tst-pam_faillock-retval.c:
```diff
@@ -0,0 +1,115 @@
+	ASSERT_EQ(0, unlink(service_file));
+	pamh = NULL;
+
+	ASSERT_LT(0,asprintf(&filename, "%s/%s", cwd, user_name));
+	ASSERT_EQ(0,unlink(filename));
+	free(filename);
```
Yet again. filename doesn't have to be an absolute path, I suppose you can just do
```
ASSERT_EQ(0, unlink(user_name));
```
instead.

ldv-alt reviewed (changes requested):
Stefan, thanks a lot for contributing a test for pam_faillock.
I think we need a test of this kind for every module, not just for a dozen of them.

I wonder, however, why you decided to bundle the test into the vendordir PR - these two changes (the test and all the rest) are unrelated.

I would prefer if this change was split into two commits: the first one is the new test, and the second one is the vendordir support.

And the last but not least is the commit message.  It's too brief.
Could you copy the description of this PR into the commit message, please?

schubi2 commented:
> I would prefer if this change was split into two commits: the first one is the new test, and the second one is the vendordir support.
I have made an own PR for testcases:
https://github.com/linux-pam/linux-pam/pull/431/files

ldv-alt reviewed (approved):
I've tweaked wording in pam_faillock.8.xml, fixed some spaces-vs-tabs issues in the patch,
and changed the commit message a bit.

I suppose it's OK to push to master.

schubi2 commented:
> I've tweaked wording in pam_faillock.8.xml, fixed some spaces-vs-tabs issues in the patch, and changed the commit message a bit.
> 
> I suppose it's OK to push to master.

Thank you !
